doc/manual/src/store/index.md

```diff
@@ -0,0 +1,84 @@
+# Nix Store
+
+A Nix store is a collection of *store objects* with references between them.
+It supports operations to manipulate that collection.
+
+## Store Object
+
+A store object consists of
+
+  - A [file system object] as data
+  - A set of [store paths](@docroot@/glossary.md#gloss-store-path) as references to other store objects
+
+[file system object]: ./file-system-object.md
+
+Store objects are [immutable]:
+Once created, they do not change until they are deleted.
+
+[immutable]: https://en.wikipedia.org/wiki/Immutable_object
+
+## Operations
+
+A Nix store can *add*, *retrieve*, and *delete* store objects.
+
+As a visual aid to the abstract interface, the following diagrams read as follows:
+- Top to bottom: input and output
+- Left to right: change of state
+
+### Add
+
+Add a [file system object] and return a reference to the resulting store object.
+
+          [ file system object ]
+                    |
+                    V
+    [ store ] ---> add ----> [ store' ]
+                    |
+                    V
+              [ reference ]
+
+### Get
+
+Retrieve a store object given a reference.
+
+              [ reference ]
+                    |
+                    V
+    [ store ] ---> get
+                    |
+                    V
+             [ store object ]
+
+### Delete
+
+Delete a store object given a reference.
+This is only possible if the store object is not [reachable] from any reference still in use.
+
+[reachable]: @docroot@/glossary.md#gloss-reachable
+
+              [ reference ]
+                    |
+                    V
+    [ store ] --> delete --> [ store' ]
+
+### Build
+
+Some store types can *perform builds*, that is, create new store objects by transforming build inputs into build outputs, using build instructions specified in the given reference to a [store derivation](@docroot@/glossary.md#gloss-store-derivation).
+
+              [ reference ]
+                    |
+                    V
+    [ store ] --> build --(maybe)--> [ store' ]
+                             |
+                             V
+                       [ reference' ]
+
+### Collect garbage
+
+As a store keeps track of references, it can [garbage-collect] unused store objects.
+This will remove all store objects not [reachable] from garbage collector roots.
+
+    [ store ] --> collect garbage --> [ store' ]
```

Comment:
```suggestion
```

GC is an example of why the purely functional notation doesn't really work: GC is not a function from a store to store, it's a process which isn't necessarily fully serializable with other concurrent operations on the same store. Probably better to describe it in terms of its pre- and post-conditions.